File: src/python/dijkstra.py
import heapq


class Graph:
    def __init__(self, edges: list):
        self.adj = [[] for _ in range(len(edges))]
        self.dist = [99999 for _ in range(len(edges))]
        self.add_edges(edges)

    def add_edges(self, edges: list) -> None:
        for i in range(len(edges)):
            for j in range(len(edges[i])):
                self.__add_edge(i, edges[i][j])

    def __add_edge(self, u: int, v: int) -> None:
        if v[0] not in self.adj[u]:
            self.adj[u].append([v[1], v[0]])

    def _weight_between_u_and_v(self, u: int, v: int) -> float:
        for vertex in self.adj[v[1]]:
            if vertex[1] == u:
                return vertex[0]

    def dijkstra(self, start: int) -> list:
        distance = self.dist.copy()
        S = set()  # Set of explored vertices
        distance[start] = 0

        while True:
            V = set([(i, distance[i]) for i in range(len(self.adj))])
            difference_of_sets = list(V.difference(S))
            if not difference_of_sets:
                break

            heapq.heapify(difference_of_sets)
            u, distance_u = heapq.heappop(difference_of_sets)

            S.add((u, distance[u]))
            for v in self.adj[u]:
                if distance[v[1]] > distance_u + self._weight_between_u_and_v(u, v):
                    distance[v[1]] = distance_u + self._weight_between_u_and_v(u, v)

        return distance


if __name__ == "__main__":
    edges = [
        [[1, 1], [2, 0.3], [5, 0.2]],  # Neighbors of vertex 0.
        [[0, 1], [2, 0.5]],  # Neighbors of vertex 1.
        [[0, 0.3], [1, 0.5], [3, 1.5], [4, 2]],  # Neighbors of vertex 2.
        [[2, 1.5], [4, 1.3], [5, 0.8]],  # Neighbors of vertex 3.
        [[2, 2], [3, 1.3]],  # Neighbors of vertex 4.
        [[0, 0.2], [3, 0.8]],  # Neighbors of vertex 5.
    ]

    graph = Graph(edges)
    print(graph.dijkstra(0))

File: src/rust/Dijkstra.rs
use std::collections::BinaryHeap;
use std::collections::HashMap;
use std::collections::HashSet;

const INF: i32 = i32::MAX;

#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
struct Edge {
    target: usize,
    weight: i32,
}

#[derive(Debug, PartialEq, Eq, Hash, Clone)]
struct Graph {
    nodes: usize,
    edges: Vec<Vec<Edge>>,
}

impl Graph {
    // Constructor to create a new graph with a given number of nodes.
    fn new(nodes: usize) -> Self {
        Graph {
            nodes,
            edges: vec![vec![]; nodes],
        }
    }

    // Add an edge to the graph from 'from' node to 'to' node with a given weight.
    fn add_edge(&mut self, from: usize, to: usize, weight: i32) {
        self.edges[from].push(Edge { target: to, weight });
    }

    // Find the shortest path from a specified starting node using Dijkstra's algorithm.
    fn shortest_path(&self, start: usize) -> HashMap<usize, i32> {
        let mut distance: HashMap<usize, i32> = (0..self.nodes).map(|i| (i, INF)).collect();
        let mut visited: HashSet<usize> = HashSet::new();

        // Set the distance to the starting node to 0 and initialize a priority queue.
        distance.insert(start, 0);
        let mut min_heap: BinaryHeap<(i32, usize)> = BinaryHeap::new();
        min_heap.push((0, start));

        while let Some((_dist, node)) = min_heap.pop() {
            if visited.contains(&node) {
                continue;
            }

            visited.insert(node);

            // Explore neighboring nodes and update distances if a shorter path is found.
            for edge in &self.edges[node] {
                let new_dist = distance[&node] + edge.weight;
                if new_dist < distance[&edge.target] {
                    distance.insert(edge.target, new_dist);
                    min_heap.push((new_dist, edge.target));
                }
            }
        }

        distance
    }
}

fn main() {
    // Create a new graph.
    let mut graph = Graph::new(5);

    // Add edges to the graph.
    graph.add_edge(0, 1, 1);
    graph.add_edge(0, 2, 4);
    graph.add_edge(1, 2, 2);
    graph.add_edge(1, 3, 5);
    graph.add_edge(2, 3, 1);
    graph.add_edge(3, 4, 3);

    let start_node = 0;

    // Find the shortest distances from the starting node.
    let shortest_distances = graph.shortest_path(start_node);

    // Print the results.
    for (node, distance) in shortest_distances.iter() {
        println!(
            "Shortest distance from node {} to node {}: {}",
            start_node, node, distance
        );
    }
}
